File: src/search/Path.java
package search;

import java.util.ArrayList;

public class Path {
	private ArrayList<State> path;
	private int totalWeight, lastEdgeWeight;
	
	/**
	 * Create a path with one given Successor.
	 * 
	 * @param v A Successor to add
	 */
	public Path(State v) {
		path = new ArrayList<State>();
		
		path.add(v);
		totalWeight = 0;
		lastEdgeWeight = 0;
	}
	
	/**
	 * Create a path with an old path and a pair.
	 * 
	 * @param old Old path to start from
	 * @param t Successor to add on
	 * @param SuccessorWeight Weight of Successor
	 */
	public Path(Path old, State s) {
		path = new ArrayList<State>();
		
		if (old == null) {
			return;
		}
		
		//Deep copy of old into path
		ArrayList<State> oldPathway = old.getPathway();
		for (State Successor : oldPathway) {
			path.add(Successor);
		}
		
		if (s != null) {
			path.add(s);
		}
		
		if (path.size() > 1) {
			totalWeight = old.getTotalWeight() + s.getWeight();
			lastEdgeWeight = s.getWeight();
		} else {
			totalWeight = 0;
			lastEdgeWeight = 0;
		}
	}
	
	/**
	 * Return the total weight of the path.
	 * 
	 * @return Total weight of the path
	 */
	public int getTotalWeight() {
		return totalWeight;
	}
	
	/**
	 * Return the last State.
	 * 
	 * @return Last State
	 */
	public State getLastState() {
		if (path.size() > 1) {
			return path.get(path.size() - 1);
		} else if (path.size() == 1) {
			return path.get(0);
		} else {
			return null;
		}
	}
	
	/**
	 * Return the arrayList of vertices.
	 * 
	 * @return ArrayList of vertices
	 */
	public ArrayList<State> getPathway() {
		return path;
	}
	
	/**
	 * Checks whether or not a path contains the specified Successor.
	 * 
	 * @param v Successor to check for
	 * @return Whether or not that Successor is in the path
	 */
	public boolean contains(State v) {
		return path.contains(v);
	}
}

File: web/js/search/Path.js
export default class Path {
    constructor(state) {
        this._path = [state];
        this._totalWeight = 0;
        this._lastEdgeWeight = 0;
    }

    static extend(old, s) {
        const p = new Path(null);
        p._path = [];

        if (old === null) {
            return p;
        }

        const oldPathway = old.getPathway();
        for (const state of oldPathway) {
            p._path.push(state);
        }

        if (s !== null) {
            p._path.push(s);
        }

        if (p._path.length > 1) {
            p._totalWeight = old.getTotalWeight() + s.getWeight();
            p._lastEdgeWeight = s.getWeight();
        } else {
            p._totalWeight = 0;
            p._lastEdgeWeight = 0;
        }

        return p;
    }

    getTotalWeight() {
        return this._totalWeight;
    }

    getLastState() {
        if (this._path.length > 1) {
            return this._path[this._path.length - 1];
        } else if (this._path.length === 1) {
            return this._path[0];
        } else {
            return null;
        }
    }

    getPathway() {
        return this._path;
    }

    contains(state) {
        return this._path.indexOf(state) !== -1;
    }
}
